cpp/src/community/k_truss_impl.cuh

```diff
@@ -0,0 +1,951 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+#pragma once
+
+// FIXME: remove all unused imports
+#include <cugraph/algorithms.hpp>
+#include <cugraph/detail/shuffle_wrappers.hpp>
+#include <cugraph/graph_functions.hpp>
+#include <cugraph/utilities/error.hpp>
+#include <cugraph/utilities/host_scalar_comm.hpp>
+
+#include <thrust/adjacent_difference.h>
+#include <thrust/binary_search.h>
+#include <thrust/copy.h>
+#include <thrust/count.h>
+#include <thrust/distance.h>
+#include <thrust/execution_policy.h>
+#include <thrust/fill.h>
+#include <thrust/iterator/transform_iterator.h>
+#include <thrust/optional.h>
+#include <thrust/scatter.h>
+#include <thrust/sort.h>
+#include <thrust/transform.h>
+#include <thrust/tuple.h>
+
+#include <prims/edge_bucket.cuh>
+#include <prims/extract_transform_e.cuh>
+#include <prims/fill_edge_property.cuh>
+#include <prims/reduce_op.cuh>
+#include <prims/transform_e.cuh>
+#include <prims/transform_reduce_dst_nbr_intersection_of_e_endpoints_by_v.cuh>
+#include <prims/transform_reduce_e.cuh>
+#include <prims/update_edge_src_dst_property.cuh>
+
+namespace cugraph {
+
+// FIXME : This will be deleted once edge_triangle_count becomes public
+template <typename vertex_t, typename edge_t, bool store_transposed, bool multi_gpu>
+rmm::device_uvector<edge_t> edge_triangle_count(
+  raft::handle_t const& handle,
+  graph_view_t<vertex_t, edge_t, store_transposed, multi_gpu> const& graph_view,
+  raft::device_span<vertex_t> edgelist_srcs,
+  raft::device_span<vertex_t> edgelist_dsts);
+
+template <typename vertex_t, typename edge_t, typename EdgeBuffer>
+struct unroll_edge {
+  edge_t num_valid_edges{};
+  raft::device_span<edge_t> num_triangles{};
+  EdgeBuffer edge_to_unroll_first{};
+  EdgeBuffer transposed_valid_edge_first{};
+  EdgeBuffer transposed_valid_edge_last{};
+  EdgeBuffer transposed_invalid_edge_last{};
+
+  __device__ thrust::tuple<vertex_t, vertex_t> operator()(edge_t i) const
+  {
+    // edges are sorted with destination as key so reverse the edge when looking it
+    auto pair = thrust::make_tuple(thrust::get<1>(*(edge_to_unroll_first + i)),
+                                   thrust::get<0>(*(edge_to_unroll_first + i)));
+    // Find its position in either partition of the transposed edgelist
+    // An edge can be in found in either of the two partitions (valid or invalid)
+
+    auto itr = thrust::lower_bound(
+      thrust::seq, transposed_valid_edge_last, transposed_invalid_edge_last, pair);
+    auto idx = thrust::distance(transposed_valid_edge_last, itr) + num_valid_edges;
+
+    if (*itr != pair) {
+      // The edge must be in the first boundary
+      itr = thrust::lower_bound(
+        thrust::seq, transposed_valid_edge_first, transposed_valid_edge_last, pair);
+      idx = thrust::distance(transposed_valid_edge_first, itr);
+    }
+
+    assert(*itr == pair);
```

Comment:
`*itr` here can possibly access invalid memory (if itr == transposed_invalid_edge_last).

```
size_t idx{};
if (itr != transposed_invalid_edge_last && *itr == pair) {
  idx = static_cast<size_t>(thrust::distance(transposed_valid_edge_last, itr) + num_valid_edges));
}
else {
  // The edge must be in the first boundary
      itr = thrust::lower_bound(
        thrust::seq, transposed_valid_edge_first, transposed_valid_edge_last, pair);
      assert(*itr == pair);
      idx = thrust::distance(transposed_valid_edge_first, itr);
}
```

Reply:
Right  I did it in Line 151 and forgot to update it here too